crates/transport-ipc/src/lib.rs

```diff
@@ -148,19 +145,23 @@ where
         self: std::pin::Pin<&mut Self>,
         cx: &mut std::task::Context<'_>,
     ) -> std::task::Poll<Option<Self::Item>> {
-        let this = self.project();
+        use tokio_util::io::poll_read_buf;
 
-        // Deserialize any buffered items.
-        if !this.buf.is_empty() {
-            this.reader.consume(this.buf.len());
+        let mut this = self.project();
 
+        loop {
+            // try decoding from the buffer
             tracing::debug!(buf_len = this.buf.len(), "Deserializing buffered IPC data");
             let mut de = serde_json::Deserializer::from_slice(this.buf.as_ref()).into_iter();
 
             let item = de.next();
+
+            // advance the buffer
+            this.buf.advance(de.byte_offset());
+
             match item {
                 Some(Ok(response)) => {
-                    this.items.push(response);
+                    return Ready(Some(response));
```

Comment:
I guess the thing here is that you assume the read has EXACTLY ONE item in it. but that may not be the case if geth has sent multiple notifications

Reply:
oh then i guess the followup would be that the underlying stream re-wakes the task as there's more data?

Reply:
> wont this now fail to wake up again until the socket has more data?

this yields a new value, the caller is supposed to call poll_next again, streams should only register the waker if they are pending

https://docs.rs/futures/latest/futures/stream/trait.Stream.html#tymethod.poll_next

> Attempt to pull out the next value of this stream, registering the current task for wakeup if the value is not yet available, and returning None if the stream is exhausted.

the way this loop works is:

it drains the buffer by yielding deserialized values (advancing the buffer)
if there are no more items to serialize to fills the buffer by reading from the `reader` until pending

Reply:
>> wont this now fail to wake up again until the socket has more data?

> this yields a new value, the caller is supposed to call poll_next again, streams should only register the waker if they are pending

this doesn't mesh with my understanding of my understanding of wakers. AIUI the waker exists to notify the executor that there is additional work, and the executor may never poll a task that does not have a call to `Waker::wake` or `Waker::wake_by_ref`. E.g. the [example executor](https://rust-lang.github.io/async-book/02_execution/04_executor.html) would not re-poll this future until the waker in the underlying IO re-wakes the task because more data is available to read

when they say "register for waking" I'm pretty sure what they mean is "store a copy of the waker in the local state. this is how, e.g., the `tokio::time::sleep` works under the hood. it stores the waker from the context of any task passed into it.

as a side note, your code may work in prod because tokio has an optimization where it polls futures again immediately after they return ready (even if the waker is not triggered), in case there's more work. While it may work, I'm not sure it is correct

Reply:
> your code may work in prod because tokio has an optimization where it polls futures again

tokio does not poll the future directly, the user of the stream does by calling `poll_next`. This happens in a Future, or poll function, so

```rust
 while let Some(next) = stream.next().await
```

is essentially a busy loop if the stream can always yield an item, the await point is only hit once the stream returns `Pending` at which point the task goes back to the executor but the stream is supposed to have now registered a waker so the task will be polled again once it can yield another item

the implicit "guarantee" that `poll_next` registers the waker only exists if the stream is pending, per docs.  

For example: mpsc::UnboundedReceiver Stream:

https://docs.rs/futures-channel/0.3.29/src/futures_channel/mpsc/mod.rs.html#1084-1090

Reply:
> tokio does not poll the future directly, the user of the stream does by calling poll_next. This happens in a Future, or poll function

the executor deals with the top-level task which contains the stream state machine, and any futures or other streams or other structs that contain that stream, and schedules that task for execution. futures and streams use the wakers to communicate to the executor that the task should be scheduled. 

`while let Some(next) = stream.next().await`

this is a busy loop if the stream is always ready, yes. but as soon as the stream returns pending and the executor switches tasks, there is _no guarantee_ that the executor returns to task before the task's Waker is used to re-wake it. 

> the implicit "guarantee" that `poll_next` registers the waker only exists if the stream is pending, per docs.
> 
> For example: mpsc::UnboundedReceiver Stream:
> 
> https://docs.rs/futures-channel/0.3.29/src/futures_channel/mpsc/mod.rs.html#1084-1090

this is "registering" the waker by adding it to task state to be re-woken when the channel has more items, ensuring that the executor calls poll when items are available, as i described above. `ctx.waker().wake_by_ref()` is **not** registering the waker in this way, it is requesting that the task be polled again by the executor immediately

Reply:
> ctx.waker().wake_by_ref() is not registering the waker in this way, it is requesting that the task be polled again by the executor immediately

right, that's why it isn't necessary to call it when returning an item

Reply:
it is necessary if our buffer isn't empty, as we have work to do immediately, which wont get done if the task isnt rewoken

Reply:
the new poll loop has 4 exits:
1. item decoded -> return the item
2. decoding failed -> terminate
3. no item decoded, nothing read from the reading -> pending (this ensures we're getting called again as soon as there's more to read)
4. reading failed

I fail to see where we'd need to request another call.
But after writing this I think we can optimize this a bit more and skip the decoding attempt one time when we previously returned pending

Reply:
if we read every item as soon as it's written, then we don't have a problem

if the other end of the socket ever writes 2 items between polls of this stream, thne then next time the stream is polled it will fill the buffer (getting the 2 items), and then deser and return an item. At that point, we have a serialized item in the buffer. If we do not re-wake properly, then the next poll will occur when the asyncread is ready to give us more serialized items for our buffer. That wake will again result in only 1 item being read from the buffer

we need to re-wake until the buffer is empty, otherwise we risk falling significantly behind and filling our buffer